keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,705 @@
+    type NodeIP struct {
+            // IP is a node IP address
+            IP string `json:"ip"`
+
+            // Scope gives the scope of IP
+            // +optional
+            Scope NodeIPScope `json:"scope,omitempty"`
+
+            // DNSName is an optional DNS name that resolves to IP
+            // +optional
+            DNSName string `json:"dnsName,omitempty"`
+    }
+
+    // NodeIPScope describes the scope within which a Node IP is useful.
+    type NodeIPScope string
+
+    const (
+            // NodeIPUnknown indicates an IP address with unknown scope. (It is at least
+            // reachable by all other nodes within the cluster, but may or may not be
+            // accessible beyond that.)
+            NodeIPUnknown NodeIPScope = ""
+
+            // NodeIPInternal indicates an IP address that exists only within the scope of
+            // some private network (eg, an intranet or cloud VPC). The scope is at least
+            // large enough to include all other nodes within the cluster.
+            NodeIPInternal NodeIPScope = "Internal"
+
+            // NodeIPExternal indicates an IP address that exists only outside the cluster
+            // (eg, traffic sent to this address from inside the cluster may leave the
+            // cluster and then come back with the destination IP rewritten to an Internal
+            // IP address.)
+            NodeIPExternal NodeIPScope = "External"
+
+            // NodeIPPublic indicates an IP address that is expected to be directly
+            // reachable from anywhere, inside or outside the cluster.
+            NodeIPPublic NodeIPScope = "Public"
+    )
+
+Rather than having separate `NodeInternalIP`/`NodeInternalDNS` and
+`NodeExternalIP`/`NodeExternalDNS` entries, the DNS entries are now
+included along with the matching IPs.
+
+```
+<<[UNRESOLVED dns-correlation ]>>
+It seems that most clouds' APIs return IPs and DNS names via separate
+API calls, so correlating them will require extra work (eg, DNS
+lookups) and in some cases a cloud might currently be returning DNS
+names that don't correspond to any of the IPs it returns (which would
+not be possible with `NodeIPs`). Is this a problem?
+<<[/UNRESOLVED]>>
+```
+
+Additionally, the `InternalIP` vs `ExternalIP` split is now clarified
+and expanded to cover additional cases. For the most part, existing
+cloud `InternalIP` addresses would translate to `Internal` scope and
+`ExternalIP` addresses would translate to `External` scope, but IPv6
+node IPs will most likely be `Public`, and bare-metal IPs will generally
+be `Unknown`. (Private cloud providers (eg, vSphere and OpenStack) might
+also use `Unknown` scope.)
+
+```
+<<[UNRESOLVED other-info ]>>
+Do we want other information in `NodeIP`? eg, we could indicate which
+network interface is associated with each IP, or more abstractly allow
+there to be "tags" of some sort ("control plane", "storage
+network").
```

Comment:
Massive scope creep imo.  If we want this (and I think we do not), consider creating a new NodeAddress kind, and using labels/annotations.  (and then label selectors for specific purposes rather than a single "scope", etc)

Reply:
It's a thing that's been discussed before, and if we were going to solve it, then it seems like something we should think about while redesigning node IPs (even if we're not actually going to solve it now).